schematics/exceptions.py

```diff
@@ -2,9 +2,98 @@
 
 from __future__ import unicode_literals, absolute_import
 
-from collections import Sequence
+import copy
+import json
+
+from collections import Sequence, Mapping
 
 from .common import *
+from .compat import string_type
+from .datastructures import FrozenDict, FrozenList
+
+
+@str_compat
+class BaseError(Exception):
+
+    def __init__(self, message, errors=None, *args):
+        """
+        The base class for all Schematics errors.
+
+        message should be a human-readable message,
+        while errors is a machine-readable list, or dictionary.
+
+        if None is passed as the message, and error is populated,
+        the primitive representation will be serialized.
+
+        the Python logging module expects exceptions to be hashable
+        and therefore immutable. As a result, it is not possible to
+        mutate BaseError's error list or dict after initialization.
+        """
+        errors = self._freeze(errors)
+        if message is None and errors is not None:
+            message = str(self._to_primitive(errors))
+        super(BaseError, self).__init__(message)
+        self._errors = errors
+
+    @property
+    def errors(self):
+        return self._errors
+
+    @property
+    def messages(self):
+        """ an alias for errors, provided for compatibility with V1. """
+        return self._errors
+
+    def to_primitive(self):
+        """
+        converts the errors dict to a primitive representation of dicts,
+        list and strings.
+        """
+        if not hasattr(self, "_primitive"):
+            self._primitive = self._to_primitive(self._errors)
+        return self._primitive
+
+    @staticmethod
+    def _freeze(obj):
+        """ freeze common data structures to something immutable. """
+        if isinstance(obj, dict):
+            return FrozenDict(obj)
+        elif isinstance(obj, list):
+            return FrozenList(obj)
+        else:
+            return obj
+
+    @classmethod
+    def _to_primitive(cls, obj):
+        """ recursive to_primitive for basic data types. """
+        if isinstance(obj, string_type):
+            return obj
+        if isinstance(obj, Sequence):
+            return [cls._to_primitive(e) for e in obj]
+        elif isinstance(obj, Mapping):
+            return dict(
+                (k, cls._to_primitive(v)) for k, v in obj.items()
+            )
+        else:
+            return str(obj)
+
+    def __repr__(self):
+        return json.dumps(self.to_primitive())
+
+    __str__ = __repr__
```

Comment:
Do we need `repr` to be equal `str`? We can have repr return a "reconstructable" string, as that is its recommended format.